### ISSUE TITLE
Implicit dependency on lit-html and unpkg

### ISSUE THREAD

**Issue by styfle** [matthewp/haunted#2]
[Line #1](https://github.com/matthewp/haunted/blob/e38b1b268f8871626905ebec85b77838e83c10cd/haunted.js#L1) imports from a URL which seems like it could be bad practice, especially if the user wants to pin a version of `lit-html` instead of using the latest version.

Instead, maybe the API could be changed to accept the `lit-html` object or maybe just the `render()` function to maintain [low-coupling](https://en.wikipedia.org/wiki/Coupling_(computer_programming))?

It could look something like this:

```js
  import { html, render } from 'https://unpkg.com/lit-html@0.12.0/lit-html.js';
  import { Hooks } from 'https://unpkg.com/@matthewp/haunted@1.1.0/haunted.js';
  const { component, useState } = new Hooks(render);

  function Counter() {
    const [count, setCount] = useState(0);

    return html`
      <div id="count">${count}</div>
      <button type="button" @click=${() => setCount(count + 1)}>Increment</button>
    `;
  }
```

That way both `html` and `render` are coming from the same version of lit-html 😄

**Comment by matthewp** [comment]
Yeah, it's too bad we don't have [package-name-maps](https://github.com/domenic/package-name-maps) yet. Anyways, I knew that this was sort of weird, it was just for the initial version.

I was thinking of making it be `../node_modules/lit-html/lit-html.js`. That way if you are using npm locally it will use your version. That covers your case, right?

And then probably having a built version pointing to the unpkg URL.  Let me know what you think.

**Comment by styfle** [comment]
That might work.

If you do make `lit-html` an explicit dependency, then I think you would want to export `html` to avoid the problem that I am explaining (mismatched versions).

So the API could look like this:

```js
  import { component, useState, html } from 'https://unpkg.com/@matthewp/haunted/haunted.js';

  function Counter() {
    const [count, setCount] = useState(0);

    return html`
      <div id="count">${count}</div>
      <button type="button" @click=${() => setCount(count + 1)}>Increment</button>
    `;
  }
```

That way, the user doesn't really need to know about `lit-html` the same way a Node.js user doesn't really need to know about V8.

However, I still prefer the example in my original comment because it works without npm and allows other implementations of `lit-html` that follow the same interface.

**Comment by matthewp** [comment]
Exporting `html` is a great idea for people who only want the one dependency.

I prefer keeping it explicit because we *do* depend on lit-html. I don't think it's wrong to import it. But if we change it to `../node_modules/lit-html/lit-html.js` that will fix the pinning problem you pointed out. And it will make it work for local npm installs and bundler use. So double win. I'll do this either tonight or tomorrow. Thanks 😀

**Comment by justinfagnani** [comment]
If you just import from `lit-html`:

```js
 import { html, render } from 'lit-html.';
```

Then your module will work from unpkg with the `?module` query param:

```js
import { component, useState, html } from 'https://unpkg.com/@matthewp/haunted?module';
```

And it'll work with WebPack, Rollup, Parcel, `polymer serve`, TypeScript, and a lot of other tools that support node module resolution. Then it'll of course work with package name maps when they arrive.

The problem with importing `../node_modules/lit-html/lit-html.js` is that it won't work for local development of `haunted` itself, unless you've setup your dev environment that way.

```
> git clone https://github.com/matthewp/haunted.git
> cd haunted
> npm i
```

won't put `lit-html` at `../node_modules/lit-html`

**Comment by matthewp** [comment]
I don't mind slightly more inconvenient local development of haunted. I can just host my webserver inside of the `haunted` folder and it will work.  I'd prefer that over depending on special servers to use it with browser modules.

I'm willing to create as many builds as is necessary since we're just talking about replacing the specifier. For now I think I'm going to go with:

* __haunted.js__ - `https://unpkg.com/lit-html/lit-html.js`: This can be used for non-local dependencies.
* __index.js__ - `../node_modules/lit-html/lit-html.js`: For use with local haunted/lit-html installs.

Hopefully the latter will work with tooling (I would think it would). If webpack/etc. users complain we can add a build that uses `lit-html` as the specifier as well.

**Comment by stevenvachon** [comment]
If supporting other renderers such as hyperHTML is to continue, I think that there should be no dependency on lit-html; instead requiring it to be defined as part of a configuration. The current documentation causes confusion that only digging into the source code can clarify.

**Comment by stevenvachon** [comment]
@matthewp

**Comment by matthewp** [comment]
@stevenvachon you might want to take a look #81.
